File: ExampleApp/src/FPSPlayer.cs
using System;
using Prism;

namespace Example
{
    public class FPSPlayer : Behaviour
    {
        public float WalkingSpeed = 10.0F;
        public float RunSpeed = 20.0F;
        public float JumpForce = 50.0F;
        public float CameraForwardOffset = 0.2F;
        public float CameraYOffset = 0.85F;

        public float MouseSensitivity = 10.0F;

        private float m_CurrentSpeed;
        private float m_CurrentYMovement = 0.0F;

        private Vector2 m_MovementDirection = new Vector2(0.0F);
        private bool m_ShouldJump = false;

        private RigidBodyComponent m_RigidBody;
        private TransformComponent m_Transform;
        private TransformComponent m_CameraTransform;

        private Entity m_CameraEntity;

        private Vector2 m_LastMousePosition;

        private int m_CollisionCounter = 0;
        private bool Colliding => m_CollisionCounter > 0;

        private Collider[] m_Colliders = new Collider[10];

        void OnCreate()
        {
            m_Transform = GetComponent<TransformComponent>();
            m_RigidBody = GetComponent<RigidBodyComponent>();

            m_CurrentSpeed = WalkingSpeed;

            m_CameraEntity = Entity.FindEntityByTag("Camera");
            m_CameraTransform = m_CameraEntity.GetComponent<TransformComponent>();

            m_LastMousePosition = Input.GetMousePosition();

            Input.SetCursorMode(CursorMode.Locked);
        }

        void OnUpdate()
        {

            if (Input.IsKeyPressed(KeyCode.Escape) && Input.GetCursorMode() == CursorMode.Locked)
                Input.SetCursorMode(CursorMode.Normal);

            //if (Input.IsMouseButtonPressed(MouseButton.Left) && Input.GetCursorMode() == CursorMode.Normal)
            //    Input.SetCursorMode(CursorMode.Locked);

            m_CurrentSpeed = Input.IsKeyPressed(KeyCode.LeftControl) ? RunSpeed : WalkingSpeed;

            UpdateRaycasting();
            UpdateMovementInput();
            UpdateRotation();
            UpdateCameraTransform();
        }

        void OnFixedUpdate()
        {
            UpdateMovement();
        }

        private void UpdateRotation()
        {
            float ts = Time.DeltaTime;
            Vector2 currentMousePosition = Input.GetMousePosition();
            Vector2 delta = m_LastMousePosition - currentMousePosition;
            m_CurrentYMovement = delta.X * MouseSensitivity * ts;
            float xRotation = delta.Y * MouseSensitivity * ts;

            if (delta.Y != 0.0F || delta.X != 0.0F)
            {
                m_CameraTransform.Rotation += new Vector3(xRotation, m_CurrentYMovement, 0.0F);
            }

            m_CameraTransform.Rotation = new Vector3(Mathf.Clamp(m_CameraTransform.Rotation.X, -80.0F, 80.0F), m_CameraTransform.Rotation.YZ);

            m_LastMousePosition = currentMousePosition;
            m_Transform.Rotation = new Vector3(0.0f, m_CameraTransform.Rotation.Y, 0.0f);
        }

        private void UpdateMovementInput()
        {
            if (Input.IsKeyPressed(KeyCode.W))
                m_MovementDirection.Y = 1.0F;
            else if (Input.IsKeyPressed(KeyCode.S))
                m_MovementDirection.Y = -1.0F;
            else
                m_MovementDirection.Y = 0.0F;

            if (Input.IsKeyPressed(KeyCode.A))
                m_MovementDirection.X = -1.0F;
            else if (Input.IsKeyPressed(KeyCode.D))
                m_MovementDirection.X = 1.0F;
            else
                m_MovementDirection.X = 0.0F;

            m_ShouldJump = Input.IsKeyPressed(KeyCode.Space) && !m_ShouldJump;
        }

        private void UpdateRaycasting()
        {
            RaycastHit hitInfo;
            if (Input.IsKeyPressed(KeyCode.H) && Physics.Raycast(m_CameraTransform.Position + (m_CameraTransform.Transform.Forward * 5.0F), m_CameraTransform.Transform.Forward, 20.0F, out hitInfo))
            {
                var entity = Entity.FindEntityByID(hitInfo.EntityID);
                var renderer = entity?.GetComponent<MeshRendererComponent>();
                renderer?.Material.SetFloat("u_Metalness", 1.0f);
            }
            if (Input.IsKeyPressed(KeyCode.G) && Physics.Raycast(m_CameraTransform.Position + (m_CameraTransform.Transform.Forward * 5.0F), m_CameraTransform.Transform.Forward, 20.0F, out hitInfo))
            {
                var entity = Entity.FindEntityByID(hitInfo.EntityID);
                var renderer = entity?.GetComponent<MeshRendererComponent>();
                renderer?.Material.SetFloat("u_Metalness", 0.0f);
            }

            if (Input.IsKeyPressed(KeyCode.L))
            {
                int numColliders = Physics.OverlapBoxNonAlloc(m_Transform.Position, new Vector3(1.0F), m_Colliders);
                Log.Trace(numColliders);

                for (int i = 0; i < numColliders; i++)
                {
                    Collider c = m_Colliders[i];
                    Log.Trace(c);
                }
            }
        }

        private void UpdateMovement()
        {
            m_RigidBody.Rotate(new Vector3(0.0F, m_CurrentYMovement, 0.0F));

            Vector3 movement = m_CameraTransform.Transform.Right * m_MovementDirection.X + m_CameraTransform.Transform.Forward * m_MovementDirection.Y;
            movement.Normalize();
            Vector3 velocity = movement * m_CurrentSpeed;
            velocity.Y = m_RigidBody.GetLinearVelocity().Y;
            m_RigidBody.SetLinearVelocity(velocity);

            if (m_ShouldJump && Colliding)
            {
                m_RigidBody.AddForce(Vector3.Up * JumpForce, ForceMode.VelocityChange);
                m_ShouldJump = false;
            }
        }

        private void UpdateCameraTransform()
        {
            Vector3 position = m_Transform.Position + m_Transform.Transform.Forward * CameraForwardOffset;
            position.Y = m_Transform.Position.Y + CameraYOffset;
            m_CameraTransform.Position = position;
        }

        public void OnCollisionBegin(float data)
        {
            m_CollisionCounter++;
        }

        public void OnCollisionEnd(float data)
        {
            m_CollisionCounter--;
        }
    }
}

File: PrismEditor/Assets/scripts/Python/FPSPlayer.py
from Prism.Entity import Entity
from Prism.Behaviour import Behaviour
from Prism.Component import TransformComponent, RigidBodyComponent, MeshRendererComponent, ForceMode
from Prism.Physics.Collider import BoxCollider, SphereCollider, CapsuleCollider, MeshCollider
from Prism.Physics.Physics import Physics, RaycastHit
from Prism.Core.Input import Input, CursorMode, MouseButton
from Prism.Core.KeyCodes import KeyCodes
from Prism.Core.Log import Log
from Prism.Math import Vector2, Vector3
from Prism.Math.Mathf import Mathf
from Prism.Core.Time import Time


class FPSPlayer(Behaviour):
    """Python FPS 玩家控制器"""

    WalkingSpeed: float = 10.0
    RunSpeed: float = 20.0
    JumpForce: float = 50.0
    CameraForwardOffset: float = 0.2
    CameraYOffset: float = 0.85
    MouseSensitivity: float = 10.0

    _collisionCounter: int = 0

    @property
    def Colliding(self) -> bool:
        return self._collisionCounter > 0

    def OnCreate(self):
        self._transform: TransformComponent = self.GetComponent(TransformComponent)
        self._rigidBody: RigidBodyComponent = self.GetComponent(RigidBodyComponent)
        self._cameraEntity = Entity.FindEntityByTag("Camera")
        self._cameraTransform: TransformComponent = self._cameraEntity.GetComponent(TransformComponent)

        self._currentSpeed = self.WalkingSpeed

        self._movementDirection = Vector2(0.0, 0.0)
        self._shouldJump = False

        self._lastMousePosition = Input.GetMousePosition()

        Input.SetCursorMode(CursorMode.Locked)

    def OnFixedUpdate(self):
        ts = Time.DeltaTime

        if Input.IsKeyPressed(KeyCodes.Escape) and Input.GetCursorMode() == CursorMode.Locked:
            Input.SetCursorMode(CursorMode.Normal)

        # if Input.IsMouseButtonPressed(MouseButton.Left) and Input.GetCursorMode() == CursorMode.Normal:
        #     Input.SetCursorMode(CursorMode.Locked)

        self._currentSpeed = (
            self.RunSpeed if Input.IsKeyPressed(KeyCodes.LeftControl)
            else self.WalkingSpeed
        )

        self.UpdateRaycasting()
        self.UpdateMovementInput()
        self.UpdateRotation(ts)
        self.UpdateMovement()
        self.UpdateCameraTransform()

    def UpdateRotation(self, ts: float):
        currentMousePosition = Input.GetMousePosition()
        delta = self._lastMousePosition - currentMousePosition
        yRotation = delta.x * self.MouseSensitivity * ts
        xRotation = delta.y * self.MouseSensitivity * ts
        self._rigidBody.Rotate(Vector3(0.0, yRotation, 0.0))

        if delta.y != 0.0 or delta.x != 0.0:
            self._cameraTransform.Rotation += Vector3(xRotation, yRotation, 0.0)

        self._cameraTransform.Rotation = Vector3(Mathf.Clamp(self._cameraTransform.Rotation.x, -80.0, 80.0), self._cameraTransform.Rotation.YZ)

        self._lastMousePosition = currentMousePosition

    def UpdateMovementInput(self):
        if Input.IsKeyPressed(KeyCodes.W):
            self._movementDirection.y = 1.0
        elif Input.IsKeyPressed(KeyCodes.S):
            self._movementDirection.y = -1.0
        else:
            self._movementDirection.y = 0.0

        if Input.IsKeyPressed(KeyCodes.A):
            self._movementDirection.x = -1.0
        elif Input.IsKeyPressed(KeyCodes.D):
            self._movementDirection.x = 1.0
        else:
            self._movementDirection.x = 0.0

        self._shouldJump = Input.IsKeyPressed(KeyCodes.Space) and not self._shouldJump

    def UpdateRaycasting(self):
        if Input.IsKeyPressed(KeyCodes.H):
            origin = self._cameraTransform.Position + (self._cameraTransform.Transform.Forward * 5.0)
            hit = RaycastHit()
            if Physics.Raycast(origin, self._cameraTransform.Transform.Forward, 20.0, hit):
                entity = Entity.FindEntityByID(hit.EntityID)
                if entity is not None:
                    mesh: MeshRendererComponent = entity.GetComponent(MeshRendererComponent)
                    if mesh is not None and mesh.Mesh is not None:
                        mesh.GetMaterial(0).SetFloat("u_Metalness", 1.0)
        if Input.IsKeyPressed(KeyCodes.L):
            colliders = Physics.OverlapBox(self._transform.Position, Vector3(1.0, 1.0, 1.0))
            Log.Trace("Overlap count: {}", len(colliders))
            for c in colliders:
                Log.Trace("EntityID: {}", c.EntityID)
                Log.Trace("IsTrigger: {}", c.IsTrigger)
                Log.Trace("IsBox: {}", isinstance(c, BoxCollider))
                Log.Trace("IsSphere: {}", isinstance(c, SphereCollider))
                Log.Trace("IsCapsule: {}", isinstance(c, CapsuleCollider))
                Log.Trace("IsMesh: {}", isinstance(c, MeshCollider))

    def UpdateMovement(self):
        movement = self._cameraTransform.Transform.Right * self._movementDirection.x + self._cameraTransform.Transform.Forward * self._movementDirection.y
        movement.Normalize()
        velocity = movement * self._currentSpeed
        velocity.y = self._rigidBody.GetLinearVelocity().y
        self._rigidBody.SetLinearVelocity(velocity)

        if self._shouldJump and self.Colliding:
            self._rigidBody.AddForce(Vector3.Up * self.JumpForce, ForceMode.Impulse)
            self._shouldJump = False

    def UpdateCameraTransform(self):
        position = self._transform.Position + self._transform.Transform.Forward * self.CameraForwardOffset
        position.y = self._transform.Position.y + self.CameraYOffset
        self._cameraTransform.Position = position

    def OnCollisionBegin(self, collision_id: float):
        self._collisionCounter += 1

    def OnCollisionEnd(self, collision_id: float):
        self._collisionCounter -= 1
